### ISSUE TITLE
move FailWith Field generic to trait type parameters

### ISSUE THREAD

**Issue by 0lejk4** [jap-company/fields#18]
Currently all FailWith is declared like this
```scala 
trait FailWithInvalid[E] { def invalid[A](field: Field[A]): E }
```
In theory they could be encoded like this:
```scala
trait FailWithInvalid[E, P] { def invalid(field: Field[P]): E }
```
So for different Field types there could be custom FailWithInstance and would be cool if it would fallback to some default one that does not care about P type.

**Comment by 0lejk4** [comment]
Seems like this is working when used in pair with https://typelevel.org/cats/guidelines.html#implicit-instance-priority

**Comment by 0lejk4** [comment]
Fixed in 0.4.x
